s2and/data.py

```diff
@@ -1609,15 +1616,63 @@ def pair_sampling_to_store(
-                # blockwise_pairs: Dict[str, List[Tuple[str, str, Union[int, float]]]] = {}
-                # for k in blockwise_sig_pairs.keys():
-                #     block_sample_size = int(sample_size * len(blockwise_sig_pairs[k])/len(possible))
-                #     samples = random_sampling(blockwise_sig_pairs[k], block_sample_size, self.random_seed)
-                #     if(len(samples)>0):
-                #         blockwise_pairs[k] = samples
-                return blockwise_sig_pairs, blockwise_cluster_ids
+                # Subsampling, while maintaining transitivity
+                sample_size = min(len(possible), sample_size)
+                pairs_ids = random_sampling(possible, sample_size, self.random_seed)
+                # Construct hashset to make queries faster
+                subsample_id_set = set()
+                for tuple in pairs_ids:
+                    id1, id2, _ = tuple
+                    subsample_id_set.add(id1)
+                    subsample_id_set.add(id2)
+
+                # if not os.path.exists("our_data_subsample.pkl"):
+                #     with open('our_data_subsample.pkl', 'wb') as f:
+                #         pickle.dump(subsample_id_set, f)
+
+                # Remove sig_pairs which are not subsampled
+                for block_id, signatures in blocks.items():
+                    sig_pairs = blockwise_sig_pairs[block_id]
+                    cluster_ids = blockwise_cluster_ids[block_id]
+                    n = len(signatures)
+                    block_len = len(sig_pairs)
+
+                    all_idxs = np.arange(0, n)
+                    sig_idxs_to_keep = []
+                    for i, s in enumerate(signatures):
+                        if (s in subsample_id_set):
+                            sig_idxs_to_keep.append(i)
+
+                    sig_idxs_to_keep = np.sort(sig_idxs_to_keep)
+                    if(sig_idxs_to_keep.size == 0):
+                        sig_idxs_to_remove = all_idxs
+                    else:
+                        sig_idxs_to_remove = np.delete(all_idxs, sig_idxs_to_keep)
+
+                    idxs_to_remove = []
+                    for midx in sig_idxs_to_remove:
+                        idxs_to_remove += self.get_indices_by_matrix_idx(midx, n)
+                    idxs_to_remove = np.sort(np.unique(idxs_to_remove))
+                    if (idxs_to_remove.size == 0):
+                        idxs_to_keep = np.arange(block_len)
+                    else:
+                        idxs_to_keep = np.delete(np.arange(block_len), idxs_to_remove)
+
+                    idxs_to_keep = np.array(idxs_to_keep, dtype=int)
+                    sig_idxs_to_keep = np.array(sig_idxs_to_keep, dtype=int)
+                    #print(type(idxs_to_keep), type(sig_idxs_to_keep), idxs_to_keep, sig_idxs_to_keep)
+                    _sig_pairs = [ele for idx, ele in enumerate(sig_pairs) if idx in idxs_to_keep]
```

Comment:
Logically the same but computationally heavier to iterate. Might become a problem for the larger datasets. 

Does setting `sig_pairs = np.array(sig_paris)` once, and then calling `_sig_pairs = sig_pairs[idxs_to_keep]` in the loop not work?

Reply:
Nopes, that changes the datatypes of each tuple and then the featurizer starts failing. I tried `list(np.array([tuple(row) for row in sig_pairs], dtype=np.dtype('U10,U10,i'))` which is the right way to preserve all datatypes in converting to and from a numpy array. Weirdly this line works on a python console (with small example array as input), but it throws an error here, maybe our python version is older or something.